uefi/src/main.rs

```diff
@@ -0,0 +1,441 @@
+
+    Some(kernel_slice)
+}
+
+/// Creates page table abstraction types for both the bootloader and kernel page tables.
+fn create_page_tables(
+    frame_allocator: &mut impl FrameAllocator<Size4KiB>,
+) -> bootloader_x86_64_common::PageTables {
+    // UEFI identity-maps all memory, so the offset between physical and virtual addresses is 0
+    let phys_offset = VirtAddr::new(0);
+
+    // copy the currently active level 4 page table, because it might be read-only
+    log::trace!("switching to new level 4 table");
+    let bootloader_page_table = {
+        let old_table = {
+            let frame = x86_64::registers::control::Cr3::read().0;
+            let ptr: *const PageTable = (phys_offset + frame.start_address().as_u64()).as_ptr();
+            unsafe { &*ptr }
+        };
+        let new_frame = frame_allocator
+            .allocate_frame()
+            .expect("Failed to allocate frame for new level 4 table");
+        let new_table: &mut PageTable = {
+            let ptr: *mut PageTable =
+                (phys_offset + new_frame.start_address().as_u64()).as_mut_ptr();
+            // create a new, empty page table
+            unsafe {
+                ptr.write(PageTable::new());
+                &mut *ptr
+            }
+        };
+
+        // copy the first entry (we don't need to access more than 512 GiB; also, some UEFI
+        // implementations seem to create an level 4 table entry 0 in all slots)
+        new_table[0] = old_table[0].clone();
+
+        // the first level 4 table entry is now identical, so we can just load the new one
+        unsafe {
+            x86_64::registers::control::Cr3::write(
+                new_frame,
+                x86_64::registers::control::Cr3Flags::empty(),
+            );
+            OffsetPageTable::new(&mut *new_table, phys_offset)
+        }
+    };
+
+    // create a new page table hierarchy for the kernel
+    let (kernel_page_table, kernel_level_4_frame) = {
+        // get an unused frame for new level 4 page table
+        let frame: PhysFrame = frame_allocator.allocate_frame().expect("no unused frames");
+        log::info!("New page table at: {:#?}", &frame);
+        // get the corresponding virtual address
+        let addr = phys_offset + frame.start_address().as_u64();
+        // initialize a new page table
+        let ptr = addr.as_mut_ptr();
+        unsafe { *ptr = PageTable::new() };
+        let level_4_table = unsafe { &mut *ptr };
+        (
+            unsafe { OffsetPageTable::new(level_4_table, phys_offset) },
+            frame,
+        )
+    };
+
+    bootloader_x86_64_common::PageTables {
+        bootloader: bootloader_page_table,
+        kernel: kernel_page_table,
+        kernel_level_4_frame,
+    }
+}
+
+fn init_logger(st: &SystemTable<Boot>, config: BootloaderConfig) -> (PhysAddr, FrameBufferInfo) {
+    let gop = st
+        .boot_services()
+        .locate_protocol::<GraphicsOutput>()
+        .expect("failed to locate gop");
```

Comment:
We shouldn't panic when there's not graphics output.

Reply:
Started working on a pr: https://github.com/Freax13/bootloader/tree/optional-framebuffer

It should work, but it needs to be tested some more. I got a little side-tracked reinvestigating https://github.com/rust-osdev/bootloader/pull/232#issuecomment-1242928162

Reply:
I opened a pr to fix this: #268.